### PR TITLE
Avoid copying dynamic slice of sampling_metadata tensors

### PR DESCRIPTION
**Pull request by mswiniarsk**
Currently when batch has changed, we are refreshing sampling metadata on device by copying only a slice of data.
This solution works good on GPU, but has negative impact on HPU by causing recompilation for each unique slice. 
Since we are copying here only a tensor of shape [batch_size], it is better to avoid recompilation in cost of copying little more data.

Other solution could be to warmup each possible test case, but in this case we would have to warmup copy of factorial of each batch_size bucket (batch_size!), which doesn't seem like a good idea.

### CODE REVIEW

File: vllm_gaudi/utils.py [hunk 1]
```diff
@@ -29,6 +29,35 @@ def hpu_backend_string():
...
+    Copy a CPU tensor to device asynchronously by creating an empty target
+    tensor and copying the source data into it.
+    """
+    assert source.device.type == 'cpu', \
+        "Source tensor is not present in host memory!"
+    target = torch.empty(source.shape, dtype=source.dtype, device=device)
+    target.copy_(source, non_blocking=True)
+    return target
+
+
```
[the remark this answers is not in the document]
Comment (mswiniarsk) [reply vllm_gaudi/utils.py]: Good idea, we have one function now :)

File: vllm_gaudi/utils.py [hunk 2]
```diff
@@ -29,6 +29,38 @@ def hpu_backend_string():
...
+                "Source tensor must be on CPU for asynchronous transfer"
+            target = torch.empty_like(source, device=device)
+            return target.copy_(source, non_blocking=True)
+    else:
+        # Create tensor from data and transfer to device
+        if dtype is None:
+            raise ValueError(
+                "dtype must be specified when source is not a tensor")
+        cpu_tensor = torch.tensor(source, dtype=dtype, device='cpu')
+        return cpu_tensor.to(device, non_blocking=True)
```
Comment (ksmusz) [inline vllm_gaudi/utils.py:61]: We could utilize a guard clause and reduce the number of if/else statements here and decrease the indents
```suggestion
    if isinstance(source, torch.Tensor):
        if dest_tensor is not None:
            # Copy into pre-allocated destination tensor
            return dest_tensor.copy_(source, non_blocking=True)
        # Create new device tensor and copy
        assert source.device.type == 'cpu', \
            "Source tensor must be on CPU for asynchronous transfer"
        target = torch.empty_like(source, device=device)
        return target.copy_(source, non_blocking=True)
    # Create tensor from data and transfer to device
    if dtype is None:
        raise ValueError(
            "dtype must be specified when source is not a tensor")
    cpu_tensor = torch.tensor(source, dtype=dtype, device='cpu')
    return cpu_tensor.to(device, non_blocking=True)
```

**Review by ksmusz — APPROVED** [review]
